eclipse/che-operator
# Add console link

### What does this PR do?
It adds creating of ConsoleLink object if the corresponding API is available onto cluster.

### What issues does this PR fix or reference?
This PR fixes https://github.com/eclipse/che/issues/14203

### Is it tested?
It's not fully tested yet but here you can find tested cases:

* OpenShift 4.2 (crc 4.1.11 on linux with consolelink CRD registered and console 4.2 used): TLS + Self-signed cert
Console link is created or update depending if it was created before
![Screenshot_20190904_134436](https://user-images.githubusercontent.com/5887312/64248590-34983380-cf1a-11e9-8fdb-f337168217a9.png)

<details>
<summary>Logs</summary>

![Screenshot_20190904_132853](https://user-images.githubusercontent.com/5887312/64247762-3234da00-cf18-11e9-84d9-e0bde331f597.png)

</details>
  
* OpenShift 4.2 (crc 4.1.11 on linux with consolelink CRD registered and console 4.2 used): HTTP
Console link is not created
<details>
<summary>Logs</summary>

![Screenshot_20190904_133341](https://user-images.githubusercontent.com/5887312/64248011-c56e0f80-cf18-11e9-9a18-d2256d2ea681.png)

</details>


* OpenShift 4.1(crc 4.1.11 on linux)
Nothing failed, che-operator prints debug message that `consolelinks` is not supported by cluster
<details>
<summary>Logs</summary>
![Screenshot_20190904_145325](https://user-images.githubusercontent.com/5887312/64252514-d8d2a800-cf23-11e9-9aa8-90e05d14c61d.png)

</details>


* OpenShift 3.11 (minishift v1.34.1 on linux)
Nothing failed, che-operator prints debug message that `consolelinks` is not supported by cluster
<details>
<summary>Logs</summary>

![Screenshot_20190903_170055](https://user-images.githubusercontent.com/5887312/64179825-74084680-ce6c-11e9-93ad-6b108c00ab3e.png)

</details>

* Kubernetes v1.15.2 (minikube v1.3.1 on linux)
Nothing failed, che-operator prints debug message that `consolelinks` is not supported by cluster
<details>
<summary>Logs</summary>

![Screenshot_20190903_162640](https://user-images.githubusercontent.com/5887312/64177433-2689da80-ce68-11e9-874e-03f0c322b231.png)

</details>

### How to test it?
Operator image to test here: `sleshchenko/che-operator:consoleLink`
To use the image apply following patch and run `./deploy.sh`:
```
diff --git a/deploy/operator.yaml b/deploy/operator.yaml
index f7c9070..3b1810e 100644
--- a/deploy/operator.yaml
+++ b/deploy/operator.yaml
@@ -24,7 +24,7 @@ spec:
     spec:
       containers:
         - name: che-operator
-          image: quay.io/eclipse/che-operator:nightly
+          image: sleshchenko/che-operator:consoleLink
           ports:
           - containerPort: 60000
             name: metrics
```

Please note that consolelink supports just `https` (https://github.com/openshift/console-operator/blob/master/manifests/00-crd-extension-console-link.yaml#L52) so you have to set your deployment to use https. Applying this patch and running `./deploy.sh` should be enough.
